File: packages/core/web/tests/throttle.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { throttle } from "../src";

describe("throttle", () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it("executes the function immediately on first call", () => {
    const fn = vi.fn();
    const throttled = throttle(fn, 200);
    throttled();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it("ignores subsequent calls within the wait period", () => {
    const fn = vi.fn();
    const throttled = throttle(fn, 200);
    throttled();
    throttled();
    throttled();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it("allows execution again after wait period elapses", () => {
    const fn = vi.fn();
    const throttled = throttle(fn, 200);
    throttled();
    expect(fn).toHaveBeenCalledTimes(1);
    vi.advanceTimersByTime(200);
    throttled();
    expect(fn).toHaveBeenCalledTimes(2);
  });

  it("passes arguments correctly", () => {
    const fn = vi.fn();
    const throttled = throttle(fn, 200);
    throttled("hello", 42);
    expect(fn).toHaveBeenCalledWith("hello", 42);
  });

  it("cancel() resets state so next call executes immediately", () => {
    const fn = vi.fn();
    const throttled = throttle(fn, 200);
    throttled();
    expect(fn).toHaveBeenCalledTimes(1);
    throttled.cancel();
    throttled();
    expect(fn).toHaveBeenCalledTimes(2);
  });

  it("cancel() is safe to call when no call is pending", () => {
    const fn = vi.fn();
    const throttled = throttle(fn, 200);
    expect(() => throttled.cancel()).not.toThrow();
  });

  it("throws Error if wait <= 0", () => {
    expect(() => throttle(() => {}, 0)).toThrow(Error);
  });

  it("throws Error if wait is negative", () => {
    expect(() => throttle(() => {}, -100)).toThrow(Error);
  });
});

File: packages/core/flutter/test/throttle_test.dart
import 'package:flutter_test/flutter_test.dart';
import '../lib/src/throttle.dart';

void main() {
  group('throttle', () {
    test('executes the function immediately on first call', () async {
      int count = 0;
      final throttled = throttle<void>((_) => count++, const Duration(milliseconds: 200));
      throttled(null);
      expect(count, 1);
    });

    test('ignores subsequent calls within the wait period', () async {
      int count = 0;
      final throttled = throttle<void>((_) => count++, const Duration(milliseconds: 200));
      throttled(null);
      throttled(null);
      throttled(null);
      expect(count, 1);
    });

    test('allows execution again after wait period elapses', () async {
      int count = 0;
      final throttled = throttle<void>((_) => count++, const Duration(milliseconds: 50));
      throttled(null);
      expect(count, 1);
      await Future.delayed(const Duration(milliseconds: 60));
      throttled(null);
      expect(count, 2);
    });

    test('passes arguments correctly', () async {
      String? received;
      final throttled = throttle<String>((v) => received = v, const Duration(milliseconds: 200));
      throttled('hello');
      expect(received, 'hello');
    });

    test('cancel() resets state so next call executes immediately', () async {
      int count = 0;
      final throttled = throttle<void>((_) => count++, const Duration(milliseconds: 200));
      throttled(null);
      expect(count, 1);
      throttled.cancel();
      throttled(null);
      expect(count, 2);
    });

    test('cancel() is safe to call when no call is pending', () {
      final throttled = throttle<void>((_) {}, const Duration(milliseconds: 200));
      expect(() => throttled.cancel(), returnsNormally);
    });

    test('throws ArgumentError if wait is Duration.zero', () {
      expect(
        () => throttle<void>((_) {}, Duration.zero),
        throwsArgumentError,
      );
    });

    test('throws ArgumentError if wait is negative', () {
      expect(
        () => throttle<void>((_) {}, const Duration(milliseconds: -100)),
        throwsArgumentError,
      );
    });
  });
}
